nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/windowsevent/WindowsEventLogRecordReader.java

```diff
@@ -0,0 +1,626 @@
+import javax.xml.stream.events.Characters;
+import javax.xml.stream.events.StartElement;
+import javax.xml.stream.events.XMLEvent;
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.text.DateFormat;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+public class WindowsEventLogRecordReader implements RecordReader {
+
+    private final ComponentLog logger;
+    private final RecordSchema schema;
+    private boolean isArray = false;
+    private XMLEventReader xmlEventReader;
+
+    private StartElement currentRecordStartTag;
+    // Utility used to infer <Data> tag data types
+    private final XmlSchemaInference xmlSchemaInference;
+
+    private final Supplier<DateFormat> LAZY_DATE_FORMAT;
+    private final Supplier<DateFormat> LAZY_TIME_FORMAT;
+    private final Supplier<DateFormat> LAZY_TIMESTAMP_FORMAT;
+
+
+    public WindowsEventLogRecordReader(InputStream in, final String dateFormat, final String timeFormat, final String timestampFormat, ComponentLog logger)
+            throws IOException, MalformedRecordException {
+
+        this.logger = logger;
+
+        final DateFormat df = dateFormat == null ? null : DataTypeUtils.getDateFormat(dateFormat);
+        final DateFormat tf = timeFormat == null ? null : DataTypeUtils.getDateFormat(timeFormat);
+        final DateFormat tsf = timestampFormat == null ? null : DataTypeUtils.getDateFormat(timestampFormat);
+
+        LAZY_DATE_FORMAT = () -> df;
+        LAZY_TIME_FORMAT = () -> tf;
+        LAZY_TIMESTAMP_FORMAT = () -> tsf;
+
+        FilterInputStream inputStream;
+        try {
+            final XMLInputFactory xmlInputFactory = XMLInputFactory.newInstance();
+            // Avoid XXE Vulnerabilities
+            xmlInputFactory.setProperty(XMLInputFactory.SUPPORT_DTD, false);
+            xmlInputFactory.setProperty("javax.xml.stream.isSupportingExternalEntities", false);
+
+            inputStream = new NonCloseableInputStream(in);
+            inputStream.mark(Integer.MAX_VALUE);
+            xmlEventReader = xmlInputFactory.createXMLEventReader(inputStream);
+            xmlSchemaInference = new XmlSchemaInference(new TimeValueInference(dateFormat, timeFormat, timestampFormat));
+
+            // Do a streaming pass through the input looking for <Data> tags, then reset the input
+            schema = determineSchema();
+
+            // Restart the XML event stream and advance to the first Event tag
+            inputStream.reset();
+            xmlEventReader = xmlInputFactory.createXMLEventReader(inputStream);
+            if (isArray) {
+                skipToNextStartTag();
+            }
+            setNextRecordStartTag();
+        } catch (XMLStreamException e) {
+            throw new MalformedRecordException("Could not parse XML", e);
```

Comment:
I will look into that. Since the format is so specific and from a particular source, I wouldn't expect errors during parsing except for "this isn't an event log file", or "this isn't XML at all", etc. But I will look at breaking things up where it makes sense.